source/slang/check.cpp

```diff
@@ -9644,16 +9667,433 @@ namespace Slang
+            return false;
+    }
+
+    return true;
+}
+
+static bool validateThisTypeSubstitutionsMatch(
+    DiagnosticSink*             /*sink*/,
+    ThisTypeSubstitution*       /*left*/,
+    ThisTypeSubstitution*       /*right*/,
+    StructuralTypeMatchStack*   /*stack*/)
+{
+    // TODO: actual checking.
+    return true;
+}
+
+static bool validateSpecializationsMatch(
+    DiagnosticSink*             sink,
+    SubstitutionSet             left,
+    SubstitutionSet             right,
+    StructuralTypeMatchStack*   stack)
+{
+    auto ll = left.substitutions;
+    auto rr = right.substitutions;
+    for(;;)
+    {
+        // Skip any global generic substitutions.
+        if(auto leftGlobalGeneric = as<GlobalGenericParamSubstitution>(ll))
+        {
+            ll = leftGlobalGeneric->outer;
+            continue;
+        }
+        if(auto rightGlobalGeneric = as<GlobalGenericParamSubstitution>(rr))
+        {
+            rr = rightGlobalGeneric->outer;
+            continue;
+        }
+
+        // If either ran out, then we expect both to have run out.
+        if(!ll || !rr)
+            return !ll && !rr;
+
+        auto leftSubst = ll;
+        auto rightSubst = rr;
+
+        ll = ll->outer;
+        rr = rr->outer;
+
+        if(auto leftGeneric = as<GenericSubstitution>(leftSubst))
+        {
+            if(auto rightGeneric = as<GenericSubstitution>(rightSubst))
+            {
+                if(validateGenericSubstitutionsMatch(sink, leftGeneric, rightGeneric, stack))
+                {
+                    continue;
+                }
+            }
+        }
+        else if(auto leftThisType = as<ThisTypeSubstitution>(leftSubst))
+        {
+            if(auto rightThisType = as<ThisTypeSubstitution>(rightSubst))
+            {
+                if(validateThisTypeSubstitutionsMatch(sink, leftThisType, rightThisType, stack))
+                {
+                    continue;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    return true;
+}
+
+// Determine if two types "match" for the purposes of `cbuffer` layout rules.
+//
+static bool validateTypesMatch(
+    DiagnosticSink*             sink,
+    Type*                       left,
+    Type*                       right,
+    StructuralTypeMatchStack*   stack)
+{
+    if(left->Equals(right))
+        return true;
+
+    // It is possible that the types don't match exactly, but
+    // they *do* match structurally.
+
+    // Note: the following code will lead to infinite recursion if there
+    // are ever recursive types. We'd need a more refined system to
+    // cache the matches we've already found.
+
+    if( auto leftDeclRefType = as<DeclRefType>(left) )
+    {
+        if( auto rightDeclRefType = as<DeclRefType>(right) )
+        {
+            // Are they references to matching decl refs?
+            auto leftDeclRef = leftDeclRefType->declRef;
+            auto rightDeclRef = rightDeclRefType->declRef;
+
```

Comment:
This is just an overall observation, that the matching is pretty convoluted - not least because it intertwines matching with traversal. That perhaps it could be simplified (and perhaps provide some useful flexibility), if there was a function that would 'flatten' a level of the definition such that the recursing down/matching becomes comparing, flattening compare etc. Ie the comparison part is pulled out. Also adds a place to memoize sameness quite easily (without having to do within each comparison function and have spread across code). 
This may not be easy for a variety of reasons, such as getting homogeneity of flattened elements - but I thought I'd throw it out there.

Reply:
The high-order bit is again that most of this code needs to be thrown out anyway. :)

The second-order bit is that when we *do* find a mismatch, we need to be able to report it to the user in terms they can follow. So imagine we have something like:

```hlsl
// A.slang
struct X { float f; }
struct Y { X x; }
cbuffer Z { Y y; }

// B.slang
struct X { int f; } // oh no! different field type!
struct Y { X x; }
cbuffer Z { Y y; }
```

We want to be able to tell the user that `Z` is defined differently between the two files, because the `Z.Y.X.f` field has different types (`float` vs. `int`), but being able to produce an error message that references all the steps along the chain requires remembering how we got to the `f` field in the first place.

The current recursive logic is using the `StructuralMatchStack` entries as a way to do that tracking, by threading a data structure up through the call stack (yes, I am aware that I use that technique a lot...). If we replaced the recursive logic with something iterative then it seems like we'd need to represent the "backtrace" for a leaf field in some other way, and the obvious ways I can think of would all involve heap allocation (maybe not a real cause for concern in practice).

Anyway, that's the other main reason for the recursive formulation, but I concede that we could do something cleaner if we wanted to keep this code long term (and I think we don't...).

Another note is that all of this is checking for the kinds of things that C++ sweeps under the rug with the One Definition Rule (ODR), so a reasonable future alternative is to say that "parameters with the same name are the same parameter, and their definitions *must* match" but not actually give people an error message on the mismatch, and leave it as undefined behavior, like C++ compilers do with ODR violations.